### PR TITLE
Return noop function from {{fn}} helper if invoked without arguments

### PR DESCRIPTION
**Pull request by jelhan**
[Rendered](https://github.com/jelhan/rfcs/blob/argumentless-fn-helper-returns-noop-function/text/0676-argumentless-fn-helper-returns-noop-function.md)

Some preliminary discussion has happened in #660.

### CODE REVIEW

File: text/0000-argumentless-fn-helper-returns-noop-function.md [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+package or expose a noop function from backing JavaScript class.
+
+A noop function is mainly required in components that have an optional
+argument, which should be either passed down to another component as an action
+handler or used with an event listener registered through `{{on}}` modifier.
+The value of the optional argument is `undefined` if consumer does not set it
+but the component may and the `{{on}}` modifier do not support `undefined` but
+require a function.
+
+Let's have a look at two example use cases to illustrate the issue:
```
Comment (jrjohnson) [inline text/0000-argumentless-fn-helper-returns-noop-function.md:41]: Another place I frequently ready for `noop` is in integration tests when invoking a component with a passed action that isn't needed in the context of the specific test:
```
await render(hbs`<VocabulariesExpanded
  @vocabularies={{this.vocabularies}}
  @collapse={{noop}}
/>`);
```

Comment (jelhan) [reply text/0000-argumentless-fn-helper-returns-noop-function.md:41]: I get what you mean. But I also experienced that there is value in being more explicit about if the function should be called or if not. You may miss some bugs if you don't assert that a function is *not* executed if it should not be executed.

Comment (jrjohnson) [reply text/0000-argumentless-fn-helper-returns-noop-function.md:41]: Makes sense. I tent to treat that as a unique unit in an isolated test. Either way, I like empty `fn` a lot.

**Comment by jelhan** [comment]
I added harmonization of public APIs of `{{array}}`, `{{hash}}` and `{{fn}}` helper as another motivation. I had lost that point in a late iteration on the draft before publishing it. Additionally I added providing a `{{noop}}` helper as an alternative. This was suggested by @kategengler in Discord.

**Comment by amk221** [comment]
+1 for this.
The ember template lint rule would have to be updated? Because you'd get `Ambiguous path 'fn' is not allowed`

**Comment by jelhan** [comment]
Added teaching section in 0fb24acb3490bb6eac286611fe7575ae76a3b4e3. Thanks to  @locks for pointing out that it's missing. To be honest I assumed that I had written it.

**Comment by jelhan** [comment]
`lint-frontmatter` and `check-filename` CI jobs are failing with an infrastructure issue, which doesn't seem to be related to this merge request:

> Error: The head commit for this pull_request event is not ahead of the base commit. Please submit an issue on this action's GitHub repo.

**Comment by pzuraq** [comment]
We discussed this in the core team meeting today, and the outcome was that with dynamic modifiers landing imminently, one of the motivating use cases for this feature, using `{{on}}` dynamically, will be solved in a way that is potentially better overall:

```js
<button {{(if @onClick on) "click" @onClick}}>
  {{yield}}
</button>
```

This is similar to the method that was mentioned in the RFC, but without the need for additional syntax (i.e. defining the semantics of `if` in modifier position). We also realized there were other motivating examples, and the consensus was that it would be good to flesh these out a bit more and also take some time to reflect on the new syntax and whether or not `{{fn}}` defaulting to no-op continues to make sense now that we (almost!) have it.

**Comment by wagenet** [comment]
Per @pzuraq's comment it sounds like this may not be needed anymore. Also template imports will change this some too. If I'm mistaken, please let me know and I can reopen it and work to get it moving forwards again.

**Pull request closed without merging**